File: kotlin-java/src/main/kotlin/programmers/lv3/여행경로/여행경로.kt
package programmers.lv3.여행경로

lateinit var visited: Array<Boolean>
lateinit var candiRoute: ArrayList<String>

fun solution(tickets: Array<Array<String>>): Array<String> {
    visited = Array(tickets.size) { false }
    candiRoute = ArrayList<String>()

    dfs(tickets, "ICN", 0, "ICN")

    candiRoute.sort()

    return candiRoute[0].split(" ").toTypedArray()
}

fun dfs(tickets:Array<Array<String>>, route:String, cnt: Int, currentCity: String) {

    if(tickets.size == cnt) {
        candiRoute.add(route)
        return
    }

    for(i in 0 until tickets.size) {
        if(!visited[i] && tickets[i][0] == currentCity) {
            visited[i] = true
            dfs(tickets, "${route} ${tickets[i][1]}", cnt + 1, tickets[i][1])
            visited[i] = false
        }
    }
}

File: Python/programmers/lv3/여행경로.py
def solution(tickets):
    complete_paths = []
    dfs(tickets, "ICN", ["ICN"], 0, complete_paths, [False] * len(tickets))
    complete_paths.sort()
    return complete_paths[0]


def dfs(tickets, cur, path, depth, complete_paths, visited):
    if depth == len(tickets):
        complete_paths.append(path)
        return

    for index, ticket in enumerate(tickets):
        (start, end) = ticket
        if visited[index] == False and cur == start:
            visited[index] = True
            dfs(tickets, end, path + [end], depth + 1, complete_paths, visited)
            visited[index] = False
